feat: Processes list pausing

vandabbin opened the pull request:
(Well kind of. It addresses the ability to pause the process list and examine it as is mentioned in a comment on this issue. I think the background update turned off in main menu is probably sufficient for a full pause with no collection)

This change sets a keybinding 'F' that pauses the processes list. It doesn't stop collection.
It stops new processes from being added and dead processes from being removed. It also disables sorting except when the sorting mode is changed or when switching between normal and tree view modes. While paused the user can still scroll through the processes list. Detailed view can still be opened and it displays current information even when the process is dead. When pausing is enabled it displays a banner on the last line of the processes list.

I think this feature is very useful. It really helps process navigation when they aren't all jumping around. 
It allows you to capture a moment in time and see what was running.

The BSDs need testing still. I have tested a bit in a FreeBSD VM. However it doesn't seem like the BSD builds are in great shape. I noticed several bugs in the Process Box on FreeBSD that exist in v1.4.5 (before any of my changes). I will be creating issues for them soon. It is possible they are because of the very minimal VM i was testing in though. As far as I could tell my changes were working on BSD though I may have missed some edge cases. 

I've been testing a lot on Linux and macOS and I think I have ironed out any issues on those platforms. If I missed anything though feedback would be great.

Screenshot of a paused process list with open detailed view (Changed to reflect change to pause banner text)
<img width="1932" height="1214" alt="Screenshot_20251024_104440" src="https://github.com/user-attachments/assets/221ca99e-4483-43e2-a18d-81587be90ec1" />

Video example of pausing (Pause banner text has been changed but isn't shown in this video)

https://github.com/user-attachments/assets/a0b61b93-a88c-4310-bded-08cd354c26c3

Perhaps there is situations where persistence might be useful instead in which case maybe an additional feature should be added to allow new processes to appear but keep dead processes in the list. This could be a toggle option in the options menu. Not part of this PR though unless consensus is it should be added.

vandabbin commented on src/linux/btop_collect.cpp:
```diff
@@ -3105,9 +3109,17 @@ namespace Proc {
+				for (auto& r : current_procs) {
+					if (rng::find(found, r.pid) == found.end()) r.cpu_p = 0;
+				}
+			}
```
Currently I am not resetting the memory display of dead processes to 0 like I am with CPU usage.

I think it could be useful to see how much memory a process was using if it runs and disappears really quickly. And pausing lets you do that. If you think that it should be shown accurately though I can add that in.

vandabbin commented on src/linux/btop_collect.cpp:
```diff
@@ -3105,9 +3110,17 @@ namespace Proc {
+				for (auto& r : current_procs) {
+					if (rng::find(found, r.pid) == found.end()) r.cpu_p = 0;
+				}
+			}
```
Currently I am not resetting the memory display of dead processes to 0 like I am with CPU usage

I think it could be useful to see how much memory a process was using if it runs and disappears really quickly. And pausing lets you do that. If you think that it should be shown accurately though I can add that in

vandabbin replied:
I'm not sure what bottom does as I haven't used it. But htop keeps updating the cpu bars and other info in the top and fully freezes the process list so both cpu usage and mem usage show as  they were.

however in htop you cant open a detailed view and see the `DEAD` status so you don't have any idea at all about whether the process is still  running or not.

The primary reason I decided to reset the cpu usage of dead processes was that for some reason when testing with stress-ng as a quick running process if i paused it and then the process died. If i then switched to tree mode while still  paused the tree sort was broken and the stress-ng process appeared at the top above its parents in cpu direct sort mode (with cpu usage of dead processes reset however, the tree sort works properly and the stress-ng process was properly under its parents in the tree. 

In this screenshot both stress-ng processes are dead. Htop also paused the elapsed time of the processes while paused which this PR for btop does not do until they are dead. (if you open detailed view of a dead process it will  show accurate elapsed time
<img width="1928" height="1212" alt="Screenshot_20251025_064900" src="https://github.com/user-attachments/assets/decc626f-daef-44ca-a71f-e3a88dc18b7b" />

vandabbin commented on src/btop_draw.cpp:
```diff
@@ -1965,10 +1968,16 @@ namespace Proc {
 		while (lc++ < height - 5) out += Mv::to(y+lc+1, x+1) + string(width - 2, ' ');
+		if (pause_proc_list) {
+			out += Mv::to(y + height - 2, x + 1) + Theme::c("proc_pause_bg") + string(width - 2, ' ')
+				+ Mv::to(y + height - 2, x + ((width - 24) / 2)) + Theme::c("title") + Fx::b + "Processes List Paused"
```
Excellent thanks for that!

I will make that change now!

deckstose commented on src/freebsd/btop_collect.cpp:
```diff
@@ -1128,7 +1132,8 @@ namespace Proc {
 				auto find_old = rng::find(current_procs, pid, &proc_info::pid);
-				if (find_old == current_procs.end()) {
+				//? Only add new processes if not paused
+				if (find_old == current_procs.end() and not pause_proc_list) {
```
```suggestion
				if (find_old == current_procs.end() && !pause_proc_list) {
```

as in the other PR.

deckstose commented on src/freebsd/btop_collect.cpp:
```diff
@@ -1186,9 +1191,17 @@ namespace Proc {
-			current_procs.erase(eraser.begin(), eraser.end());
+			//? Clear dead processes from current_procs if not paused
+			if (not pause_proc_list) {
+				auto eraser = rng::remove_if(current_procs, [&](const auto &element) { return not v_contains(found, element.pid); });
```
```suggestion
				auto eraser = rng::remove_if(current_procs, [&](const auto& element) { return not v_contains(found, element.pid); });
```

Left aligned reference symbols, please.

vandabbin replied:
> Left aligned reference symbols, please.

I will make this change. that is a line I didn't personally write though. I just wrapped a if statement around it. but I will still do that change

deckstose commented on src/btop_draw.cpp:
```diff
@@ -1965,10 +1968,16 @@ namespace Proc {
+		if (pause_proc_list) {
+			out += Mv::to(y + height - 2, x + 1) + Theme::c("proc_pause_bg") + string(width - 2, ' ')
+				+ Mv::to(y + height - 2, x + ((width - 22) / 2)) + Theme::c("title") + Fx::b + "Process list paused"
+				+ Fx::reset; 
```
If you feel comfortable please use `fmt::format` for new draw code.

Example:

```cpp
#include <iterator>

#include <fmt/format.h>

fmt::format_to(std::back_inserter(out), "{}{}{}", Fx::bold, "Hello World!", Fx::reset);
```

Here is the [syntax spec](https://fmt.dev/12.0/syntax/).

vandabbin replied:
or like this

```
fmt::format_to(std::back_inserter(out), "{}{}{}{}{}{}Process list paused{}",
			Mv::to(y + height - 2, x + 1), Theme::c("proc_pause_bg"), string(width - 2, ' '),
			Mv::to(y + height - 2, x + ((width - 22) / 2)), Theme::c("title"),
			Fx::b, Fx::reset);
```

deckstose replied:
Exactly.

You can also get rid of the `string(width - 2, ' ')` by using the align property.

It works like

```cpp
// Means: '<' align right, '{}' is the place holder for the width variable.
// And the fill char is " " (a space)
fmt::format_to(std::back_inserter(out), "{:<{}}", " ", width - 2); 
```

vandabbin replied:
awesome! thanks!

vandabbin replied:
```
fmt::format_to(std::back_inserter(out), "{}{}{}{}{:^{}}{}", 
	Mv::to(y + height - 2, x + 1),               
	Theme::c("proc_pause_bg"), Theme::c("title"),           
	Fx::b, "Process list paused", width - 2,                
	Fx::reset);                                             
```

I like this much better

I should have spent more time reading that syntax link you sent earlier haha

vandabbin commented:
@deckstose 
I saw the issue I had linked was removed. should I create a new issue to link with this PR or is it fine as is?

aristocratos commented:
> On first sight, again please remove the `and or not` stuff.

@deckstose
https://github.com/aristocratos/btop/blob/main/CONTRIBUTING.md

`and or not` is fine, since it's the convention used when I originally wrote btop.

aristocratos commented:
> This looks really nice!

Agreed!

vandabbin commented on src/btop_input.cpp:
```diff
@@ -327,7 +327,10 @@ namespace Input {
 				}
-
+				else if (is_in(key, "F")) {
+					Config::set("pause_proc_list", Config::getB("pause_proc_list") ? false : true);
```
Oh neat! I missed that! I will use that instead

aristocratos commented on src/freebsd/btop_collect.cpp:
```diff
@@ -1076,14 +1077,17 @@ namespace Proc {
 			current_sort = sorting;
 			current_rev = reverse;
 		}
+		if (tree_mode_change) is_tree_mode = tree;
```
I disagree.

vandabbin commented:
> One small thing is the left-right alignment of the pause message.
>
> On the left the bar aligns with the border, on the right it aligns with the scroll bar. Maybe we center the bar inside the proc box by moving the offset of by one?

Ahh thanks for catching that. I moved the offset over by 1

deckstose commented:
No I was wrong. If you scroll down in the proc list the bar of the selected process has the same position and width of your previous implementation, and looks off if we change that. That's why I removed the comment.

vandabbin commented:
> No I was wrong. If you scroll down in the proc list the bar of the selected process has the same position and width of your previous implementation, and looks off if we change that. That's why I removed the comment

@deckstose 
Oh! well you were still kind of right. When I changed the text from `Processes` to `Process` I misaligned the text by 1 to the left. its fixed now haha

vandabbin commented:
@aristocratos 
@deckstose 

I have made the fixes for the issues found.

- For the tree sorting I had missed a block of code in the tree generation code that was removing parent pids of  dead processes
- When cpu usage wasn't reset and tree sorting was working properly now, if proc_aggregate was set then the cpu usage of the parent processes of a dead process would just keep increasing.

I have also added in a menu option to enable or disable the clearing of cpu usage and mem usage for dead processes

If these changes look good I will squash it down to the primary  commit. 

I haven't run into any issues with my testing.

vandabbin commented:
> > It's not inherently the problem of this PR but I don't like the code duplication in all the `btop_collect.cpp` files. Maybe we need to figure out a way of combining the code in `btop_shared.cpp`
> 
> I also found this to be a bit frustrating... but at least for the most part i can modify changes i made to all of them at once with multi-file search and replace lmfao

Haha! This just bit me. Had a single character on a line that shouldn't have been there and it got copied to all three BSDs

vandabbin commented:
So I've written the code to add a 'Persistence Mode'

It basically just uses the code I wrote for pausing. The only difference between it and pausing is that it doesn't disable sorting and new processes are added to the list. I am not sure if I should add it too this PR or not. I think it could potentially break if a new process is assigned the pid of a dead process still in the list so I think it's probably best not to add it unless I figure that out.

The benefit to the persistent mode is that if you are still having trouble catching a quick running process with pause mode (since it has to be on the list when you pause) then persistent mode is a nice way to be able to "capture new processes" and then you could switch to pause mode or something like that

vandabbin commented:
@aristocratos 
Is it ok for me to `#include <unordered_set>` in the mac and bsd versions of btop_collect? I can do it a different way if required.

aristocratos commented:
@vandabbin 
Is this completed?
You requested a review, but looks like you were still working on the PR after that.
